File: packages/invoice/src/cross-validation.test.ts
/**
 * Cross-validation test data.
 *
 * These test cases are shared between TypeScript and Go implementations.
 * Both must produce IDENTICAL results for every case.
 *
 * Run: pnpm --filter @pulpo/invoice test
 * Then compare with: go test ./invoice/ -run TestCrossValidation -v
 */
import { describe, it, expect } from "vitest";
import { calculateInvoice } from "./calculate";
import type { InvoiceLineInput, InvoiceDiscountInput } from "./types";

interface CrossTestCase {
  name: string;
  items: InvoiceLineInput[];
  discount: InvoiceDiscountInput | null;
  expected: {
    gross: string;
    net: string;
    tax: string;
    subtotal: string;
    discountTotal: string;
    count: number;
    taxBreakdownLength: number;
    items: { rowTotalGross: string; taxRateSnapshot: string }[];
  };
}

// Realistic Spanish restaurant scenarios
const cases: CrossTestCase[] = [
  {
    name: "Tapas order Canary Islands (IGIC 7%)",
    items: [
      { productId: "p1", productName: "Papas arrugadas", priceGross: "6.50", taxRate: "7", quantity: 2 },
      { productId: "p2", productName: "Cerveza", priceGross: "3.50", taxRate: "7", quantity: 4 },
      { productId: "p3", productName: "Queso asado", priceGross: "8.00", taxRate: "7", quantity: 1 },
    ],
    discount: null,
    expected: {
      gross: "35.00",
      net: "32.71",
      tax: "2.29",
      subtotal: "35.00",
      discountTotal: "0.00",
      count: 7,
      taxBreakdownLength: 1,
      items: [
        { rowTotalGross: "13.00", taxRateSnapshot: "7.00" },
        { rowTotalGross: "14.00", taxRateSnapshot: "7.00" },
        { rowTotalGross: "8.00", taxRateSnapshot: "7.00" },
      ],
    },
  },
  {
    name: "Mixed tax rates mainland (IVA 10% + 21%)",
    items: [
      { productId: "p1", productName: "Menú del día", priceGross: "12.50", taxRate: "10", quantity: 2 },
      { productId: "p2", productName: "Vino", priceGross: "18.00", taxRate: "21", quantity: 1 },
      { productId: "p3", productName: "Agua", priceGross: "2.00", taxRate: "10", quantity: 2 },
    ],
    discount: null,
    expected: {
      gross: "47.00",
      net: "41.24",
      tax: "5.76",
      subtotal: "47.00",
      discountTotal: "0.00",
      count: 5,
      taxBreakdownLength: 2,
      items: [
        { rowTotalGross: "25.00", taxRateSnapshot: "10.00" },
        { rowTotalGross: "18.00", taxRateSnapshot: "21.00" },
        { rowTotalGross: "4.00", taxRateSnapshot: "10.00" },
      ],
    },
  },
  {
    name: "10% global discount on mixed rates",
    items: [
      { productId: "p1", productName: "Pulpo a la gallega", priceGross: "14.00", taxRate: "7", quantity: 1 },
      { productId: "p2", productName: "Gin Tonic", priceGross: "9.50", taxRate: "7", quantity: 2 },
    ],
    discount: { type: "percent", value: 10 },
    expected: {
      gross: "29.70",
      net: "27.76",
      tax: "1.94",
      subtotal: "33.00",
      discountTotal: "3.30",
      count: 3,
      taxBreakdownLength: 1,
      items: [
        { rowTotalGross: "12.60", taxRateSnapshot: "7.00" },
        { rowTotalGross: "17.10", taxRateSnapshot: "7.00" },
      ],
    },
  },
  {
    name: "Fixed 5€ discount with three tax groups",
    items: [
      { productId: "p1", productName: "Pan", priceGross: "1.50", taxRate: "0", quantity: 2 },
      { productId: "p2", productName: "Ensalada", priceGross: "8.50", taxRate: "10", quantity: 1 },
      { productId: "p3", productName: "Whisky", priceGross: "12.00", taxRate: "21", quantity: 1 },
    ],
    discount: { type: "fixed", value: 5 },
    expected: {
      gross: "18.50",
      net: "16.25",
      tax: "2.25",
      subtotal: "23.50",
      discountTotal: "5.00",
      count: 4,
      taxBreakdownLength: 3,
      items: [
        { rowTotalGross: "2.36", taxRateSnapshot: "0.00" },
        { rowTotalGross: "6.69", taxRateSnapshot: "10.00" },
        { rowTotalGross: "9.45", taxRateSnapshot: "21.00" },
      ],
    },
  },
  {
    name: "Item discount + global discount combined",
    items: [
      {
        productId: "p1", productName: "Langosta", priceGross: "45.00", taxRate: "7", quantity: 1,
        discount: { type: "percent", value: 15 },
      },
      { productId: "p2", productName: "Champagne", priceGross: "35.00", taxRate: "7", quantity: 1 },
    ],
    discount: { type: "fixed", value: 3 },
    expected: {
      gross: "70.25",
      net: "65.65",
      tax: "4.60",
      subtotal: "73.25",
      discountTotal: "3.00",
      count: 2,
      taxBreakdownLength: 1,
      items: [
        { rowTotalGross: "36.68", taxRateSnapshot: "7.00" },
        { rowTotalGross: "33.57", taxRateSnapshot: "7.00" },
      ],
    },
  },
  {
    name: "Single cent item (precision edge case)",
    items: [
      { productId: "p1", productName: "Chicle", priceGross: "0.01", taxRate: "21", quantity: 3 },
    ],
    discount: null,
    expected: {
      gross: "0.03",
      net: "0.02",
      tax: "0.01",
      subtotal: "0.03",
      discountTotal: "0.00",
      count: 3,
      taxBreakdownLength: 1,
      items: [
        { rowTotalGross: "0.03", taxRateSnapshot: "21.00" },
      ],
    },
  },
  {
    name: "Large quantity with rounding",
    items: [
      { productId: "p1", productName: "Café", priceGross: "1.30", taxRate: "7", quantity: 137 },
    ],
    discount: null,
    expected: {
      gross: "178.10",
      net: "166.45",
      tax: "11.65",
      subtotal: "178.10",
      discountTotal: "0.00",
      count: 137,
      taxBreakdownLength: 1,
      items: [
        { rowTotalGross: "178.10", taxRateSnapshot: "7.00" },
      ],
    },
  },
];

describe("Cross-validation with Go", () => {
  for (const tc of cases) {
    it(tc.name, () => {
      const result = calculateInvoice(tc.items, tc.discount);

      expect(result.gross).toBe(tc.expected.gross);
      expect(result.net).toBe(tc.expected.net);
      expect(result.tax).toBe(tc.expected.tax);
      expect(result.subtotal).toBe(tc.expected.subtotal);
      expect(result.discountTotal).toBe(tc.expected.discountTotal);
      expect(result.count).toBe(tc.expected.count);
      expect(result.taxBreakdown).toHaveLength(tc.expected.taxBreakdownLength);

      for (let i = 0; i < tc.expected.items.length; i++) {
        expect(result.items[i]!.rowTotalGross).toBe(tc.expected.items[i]!.rowTotalGross);
        expect(result.items[i]!.taxRateSnapshot).toBe(tc.expected.items[i]!.taxRateSnapshot);
      }

      // Invariant: net + tax === gross
      const net = parseFloat(result.net);
      const tax = parseFloat(result.tax);
      const gross = parseFloat(result.gross);
      expect(net + tax).toBeCloseTo(gross, 2);
    });
  }
});

File: apps/pulpo-app/invoice/cross_validation_test.go
package invoice

import (
	"testing"

	"github.com/shopspring/decimal"
)

// Cross-validation test cases — identical to packages/invoice/src/cross-validation.test.ts.
// Both implementations MUST produce the same results.

type crossTestCase struct {
	name     string
	items    []LineInput
	discount *DiscountInput
	expected struct {
		gross              string
		net                string
		tax                string
		subtotal           string
		discountTotal      string
		count              int
		taxBreakdownLength int
		items              []struct {
			rowTotalGross   string
			taxRateSnapshot string
		}
	}
}

func d(s string) decimal.Decimal { return decimal.RequireFromString(s) }

func getCrossValidationCases() []crossTestCase {
	cases := []crossTestCase{
		{
			name: "Tapas order Canary Islands (IGIC 7%)",
			items: []LineInput{
				{ProductID: "p1", ProductName: "Papas arrugadas", PriceGross: d("6.50"), TaxRate: d("7"), Quantity: 2},
				{ProductID: "p2", ProductName: "Cerveza", PriceGross: d("3.50"), TaxRate: d("7"), Quantity: 4},
				{ProductID: "p3", ProductName: "Queso asado", PriceGross: d("8.00"), TaxRate: d("7"), Quantity: 1},
			},
		},
		{
			name: "Mixed tax rates mainland (IVA 10% + 21%)",
			items: []LineInput{
				{ProductID: "p1", ProductName: "Menú del día", PriceGross: d("12.50"), TaxRate: d("10"), Quantity: 2},
				{ProductID: "p2", ProductName: "Vino", PriceGross: d("18.00"), TaxRate: d("21"), Quantity: 1},
				{ProductID: "p3", ProductName: "Agua", PriceGross: d("2.00"), TaxRate: d("10"), Quantity: 2},
			},
		},
		{
			name: "10% global discount on mixed rates",
			items: []LineInput{
				{ProductID: "p1", ProductName: "Pulpo a la gallega", PriceGross: d("14.00"), TaxRate: d("7"), Quantity: 1},
				{ProductID: "p2", ProductName: "Gin Tonic", PriceGross: d("9.50"), TaxRate: d("7"), Quantity: 2},
			},
			discount: &DiscountInput{Type: "percent", Value: d("10")},
		},
		{
			name: "Fixed 5€ discount with three tax groups",
			items: []LineInput{
				{ProductID: "p1", ProductName: "Pan", PriceGross: d("1.50"), TaxRate: d("0"), Quantity: 2},
				{ProductID: "p2", ProductName: "Ensalada", PriceGross: d("8.50"), TaxRate: d("10"), Quantity: 1},
				{ProductID: "p3", ProductName: "Whisky", PriceGross: d("12.00"), TaxRate: d("21"), Quantity: 1},
			},
			discount: &DiscountInput{Type: "fixed", Value: d("5")},
		},
		{
			name: "Item discount + global discount combined",
			items: []LineInput{
				{ProductID: "p1", ProductName: "Langosta", PriceGross: d("45.00"), TaxRate: d("7"), Quantity: 1,
					Discount: &DiscountInput{Type: "percent", Value: d("15")}},
				{ProductID: "p2", ProductName: "Champagne", PriceGross: d("35.00"), TaxRate: d("7"), Quantity: 1},
			},
			discount: &DiscountInput{Type: "fixed", Value: d("3")},
		},
		{
			name: "Single cent item (precision edge case)",
			items: []LineInput{
				{ProductID: "p1", ProductName: "Chicle", PriceGross: d("0.01"), TaxRate: d("21"), Quantity: 3},
			},
		},
		{
			name: "Large quantity with rounding",
			items: []LineInput{
				{ProductID: "p1", ProductName: "Café", PriceGross: d("1.30"), TaxRate: d("7"), Quantity: 137},
			},
		},
	}

	// Expected values — taken from TypeScript implementation output
	cases[0].expected.gross = "35.00"
	cases[0].expected.net = "32.71"
	cases[0].expected.tax = "2.29"
	cases[0].expected.subtotal = "35.00"
	cases[0].expected.discountTotal = "0.00"
	cases[0].expected.count = 7
	cases[0].expected.taxBreakdownLength = 1
	cases[0].expected.items = []struct{ rowTotalGross, taxRateSnapshot string }{
		{"13.00", "7.00"}, {"14.00", "7.00"}, {"8.00", "7.00"},
	}

	cases[1].expected.gross = "47.00"
	cases[1].expected.net = "41.24"
	cases[1].expected.tax = "5.76"
	cases[1].expected.subtotal = "47.00"
	cases[1].expected.discountTotal = "0.00"
	cases[1].expected.count = 5
	cases[1].expected.taxBreakdownLength = 2
	cases[1].expected.items = []struct{ rowTotalGross, taxRateSnapshot string }{
		{"25.00", "10.00"}, {"18.00", "21.00"}, {"4.00", "10.00"},
	}

	cases[2].expected.gross = "29.70"
	cases[2].expected.net = "27.76"
	cases[2].expected.tax = "1.94"
	cases[2].expected.subtotal = "33.00"
	cases[2].expected.discountTotal = "3.30"
	cases[2].expected.count = 3
	cases[2].expected.taxBreakdownLength = 1
	cases[2].expected.items = []struct{ rowTotalGross, taxRateSnapshot string }{
		{"12.60", "7.00"}, {"17.10", "7.00"},
	}

	cases[3].expected.gross = "18.50"
	cases[3].expected.net = "16.25"
	cases[3].expected.tax = "2.25"
	cases[3].expected.subtotal = "23.50"
	cases[3].expected.discountTotal = "5.00"
	cases[3].expected.count = 4
	cases[3].expected.taxBreakdownLength = 3
	cases[3].expected.items = []struct{ rowTotalGross, taxRateSnapshot string }{
		{"2.36", "0.00"}, {"6.69", "10.00"}, {"9.45", "21.00"},
	}

	cases[4].expected.gross = "70.25"
	cases[4].expected.net = "65.65"
	cases[4].expected.tax = "4.60"
	cases[4].expected.subtotal = "73.25"
	cases[4].expected.discountTotal = "3.00"
	cases[4].expected.count = 2
	cases[4].expected.taxBreakdownLength = 1
	cases[4].expected.items = []struct{ rowTotalGross, taxRateSnapshot string }{
		{"36.68", "7.00"}, {"33.57", "7.00"},
	}

	cases[5].expected.gross = "0.03"
	cases[5].expected.net = "0.02"
	cases[5].expected.tax = "0.01"
	cases[5].expected.subtotal = "0.03"
	cases[5].expected.discountTotal = "0.00"
	cases[5].expected.count = 3
	cases[5].expected.taxBreakdownLength = 1
	cases[5].expected.items = []struct{ rowTotalGross, taxRateSnapshot string }{
		{"0.03", "21.00"},
	}

	cases[6].expected.gross = "178.10"
	cases[6].expected.net = "166.45"
	cases[6].expected.tax = "11.65"
	cases[6].expected.subtotal = "178.10"
	cases[6].expected.discountTotal = "0.00"
	cases[6].expected.count = 137
	cases[6].expected.taxBreakdownLength = 1
	cases[6].expected.items = []struct{ rowTotalGross, taxRateSnapshot string }{
		{"178.10", "7.00"},
	}

	return cases
}

func TestCrossValidation(t *testing.T) {
	for _, tc := range getCrossValidationCases() {
		t.Run(tc.name, func(t *testing.T) {
			result := CalculateInvoice(tc.items, tc.discount)

			if result.Gross != tc.expected.gross {
				t.Errorf("gross: got %s, want %s", result.Gross, tc.expected.gross)
			}
			if result.Net != tc.expected.net {
				t.Errorf("net: got %s, want %s", result.Net, tc.expected.net)
			}
			if result.Tax != tc.expected.tax {
				t.Errorf("tax: got %s, want %s", result.Tax, tc.expected.tax)
			}
			if result.Subtotal != tc.expected.subtotal {
				t.Errorf("subtotal: got %s, want %s", result.Subtotal, tc.expected.subtotal)
			}
			if result.DiscountTotal != tc.expected.discountTotal {
				t.Errorf("discountTotal: got %s, want %s", result.DiscountTotal, tc.expected.discountTotal)
			}
			if result.Count != tc.expected.count {
				t.Errorf("count: got %d, want %d", result.Count, tc.expected.count)
			}
			if len(result.TaxBreakdown) != tc.expected.taxBreakdownLength {
				t.Errorf("taxBreakdown length: got %d, want %d", len(result.TaxBreakdown), tc.expected.taxBreakdownLength)
			}

			for i, exp := range tc.expected.items {
				if i >= len(result.Items) {
					t.Errorf("item %d: missing", i)
					continue
				}
				if result.Items[i].RowTotalGross != exp.rowTotalGross {
					t.Errorf("item %d rowTotalGross: got %s, want %s", i, result.Items[i].RowTotalGross, exp.rowTotalGross)
				}
				if result.Items[i].TaxRateSnapshot != exp.taxRateSnapshot {
					t.Errorf("item %d taxRateSnapshot: got %s, want %s", i, result.Items[i].TaxRateSnapshot, exp.taxRateSnapshot)
				}
			}

			// Invariant: net + tax == gross
			net := decimal.RequireFromString(result.Net)
			tax := decimal.RequireFromString(result.Tax)
			gross := decimal.RequireFromString(result.Gross)
			if !net.Add(tax).Equal(gross) {
				t.Errorf("invariant violated: %s + %s != %s", result.Net, result.Tax, result.Gross)
			}
		})
	}
}
